upstairs/src/upstairs.rs

```diff
@@ -1420,7 +1453,62 @@ impl Upstairs {
                 c.halt_io_task(ClientStopReason::Timeout);
             }
             ClientAction::Response(m) => {
-                self.on_client_message(client_id, m).await;
+                // We have received a message, so reset the timeout watchdog for
+                // this particular client.
+                self.downstairs.clients[client_id].reset_timeout();
+
+                // Defer the message if it's a (large) read that needs
+                // decryption, or there are other deferred messages in the queue
+                // (to preserve order).  Otherwise, handle it immediately.
+                if let Message::ReadResponse { responses, .. } = &m {
+                    // Any read larger than this constant should be deferred to
+                    // the worker pool; smaller reads can be processed in-thread
+                    // (since the overhead isn't worth it)
+                    const MIN_DEFER_SIZE_BYTES: u64 = 8192;
+                    let should_defer = !self.deferred_msgs.is_empty()
+                        || match responses {
+                            Ok(rs) => {
+                                // Find the number of bytes being decrypted
+                                let response_size = rs.len() as u64
+                                    * self
+                                        .ddef
+                                        .get_def()
+                                        .map(|b| b.block_size())
```

Comment:
This multiplication assumes that we only send block sized chunks - we currently do, but why not sum up the ReadResponse data lengths?

Reply:
No particular reason, other than it's O(1) instead of O(n).  I can switch if you have a preference, although I think if we switched away from block-size chunks, **many other** places would break too.

Reply:
Agreed, if we decide to support non-block-sized chunks, the existing tests will all happily tell us what to fix :)